formlibrary/views.py

```diff
@@ -249,3 +246,166 @@ def get_context_data(self, **kwargs):
         context['current_household'] = self.get_object()
         context['active'] = ['formlibrary']
         return context
+
+
+class TrainingView(ListCreateAPIView, RetrieveUpdateDestroyAPIView):
+    queryset = Training.objects.all()
+    serializer_class = TrainingSerializer
+    permission_classes = [IsAuthenticated]
+
+    def post(self, request, *args, **kwargs):
+        try:
+            request.data['program'] = request.data['program_id']
+            request.data['created_by'] = self.request.user.activity_user.id
+            return self.create(request, *args, **kwargs)
+        except Exception as e:
+            print(e)
+
+    def get_queryset(self):
+        organization = self.request.user.activity_user.organization
+        get_programs = Program.objects.all().filter(organization=organization)
+        return Training.objects.filter(program__in=get_programs)
+
+    def delete(self, request, *args, **kwargs):
+        return self.destroy(request, *args, **kwargs)
+
+
+class TrainingUpdate(UpdateView):
+    """
+    Training update
+    """
+    model = Training
+    template_name = 'formlibrary/training_form.html'
+    success_url = '/formlibrary/services_list'
+    form_class = TrainingForm
+
+    # add the request to the kwargs
+    def get_form_kwargs(self):
+        kwargs = super(TrainingUpdate, self).get_form_kwargs()
+        kwargs['request'] = self.request
+        kwargs['organization'] = self.request.user.activity_user.organization
+        return kwargs
+
+    def get_context_data(self, **kwargs):
+        context = super(TrainingUpdate, self).get_context_data(**kwargs)
+        context['current_program'] = self.get_object()
+        context['active'] = ['formlibrary']
+        return context
+
+
+class GetTrainingData(GView):
+
+    def get(self, request):
+        try:
```

Comment:
Avoid using try..except blocks. Let's remove those and see what errors we are getting and fix those errors properly.

Reply:
@nasirhjafri for me to be able to read the error from the request i used ```return JsonResponse(dict(error=str(e)))``` which made it easier to read from a json request